File: src/mecanum_bringup/scripts/camera_node.py
#!/usr/bin/env python3
# camera_node.py — publishes color camera frames using OpenCV directly
# Bypasses usb_cam driver entirely, works on any V4L2 camera
#
# Usage:
#   python3 camera_node.py
#
# Install deps if needed:
#   pip3 install opencv-python --break-system-packages

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
import cv2
import numpy as np

class CameraNode(Node):
    def __init__(self):
        super().__init__('camera_forward')

        self.declare_parameter('device_id',    1)
        self.declare_parameter('width',        640)
        self.declare_parameter('height',       480)
        self.declare_parameter('fps',          30.0)
        self.declare_parameter('topic',        '/camera/forward/image_raw')

        device_id = self.get_parameter('device_id').value
        width     = self.get_parameter('width').value
        height    = self.get_parameter('height').value
        fps       = self.get_parameter('fps').value
        topic     = self.get_parameter('topic').value

        self.pub = self.create_publisher(Image, topic, 10)

        self.cap = cv2.VideoCapture(device_id, cv2.CAP_V4L2)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH,  width)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
        self.cap.set(cv2.CAP_PROP_FPS,          fps)
        self.cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*'MJPG'))

        if not self.cap.isOpened():
            self.get_logger().error(f'Cannot open /dev/video{device_id}')
            return

        actual_w = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        actual_h = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        actual_fps = self.cap.get(cv2.CAP_PROP_FPS)
        self.get_logger().info(
            f'Camera opened: {actual_w}x{actual_h} @ {actual_fps:.1f}fps → {topic}')

        period = 1.0 / fps
        self.timer = self.create_timer(period, self.publish_frame)

    def publish_frame(self):
        ret, frame = self.cap.read()
        if not ret:
            self.get_logger().warn('Failed to capture frame')
            return

        # frame is BGR from OpenCV — convert to RGB for ROS
        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

        msg = Image()
        msg.header.stamp    = self.get_clock().now().to_msg()
        msg.header.frame_id = 'camera_forward'
        msg.height          = frame_rgb.shape[0]
        msg.width           = frame_rgb.shape[1]
        msg.encoding        = 'rgb8'
        msg.is_bigendian    = False
        msg.step            = frame_rgb.shape[1] * 3
        msg.data            = frame_rgb.tobytes()

        self.pub.publish(msg)

    def destroy_node(self):
        if self.cap.isOpened():
            self.cap.release()
        super().destroy_node()


def main():
    rclpy.init()
    node = CameraNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/mecanum_sensors/src/camera_node.cpp
// camera_node.cpp
// Captures frames from a USB/Pi camera via OpenCV (V4L2)
// and publishes them as sensor_msgs/Image on /camera/forward/image_raw
//
// Parameters:
//   device_id  (int)    — V4L2 device index, default 0 (/dev/video0)
//   width      (int)    — capture width,  default 640
//   height     (int)    — capture height, default 480
//   fps        (int)    — capture fps,    default 30
//   frame_id   (string) — TF frame,       default "camera_forward_link"

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.hpp"
#include <opencv2/opencv.hpp>

class CameraNode : public rclcpp::Node {
public:
  CameraNode() : Node("camera_node") {
    declare_parameter("device_id", 0);
    declare_parameter("width",     640);
    declare_parameter("height",    480);
    declare_parameter("fps",       30);
    declare_parameter("frame_id",  std::string("camera_forward_link"));

    int device_id = get_parameter("device_id").as_int();
    width_        = get_parameter("width").as_int();
    height_       = get_parameter("height").as_int();
    int fps       = get_parameter("fps").as_int();
    frame_id_     = get_parameter("frame_id").as_string();

    cap_.open(device_id, cv::CAP_V4L2);
    if (!cap_.isOpened()) {
      RCLCPP_ERROR(get_logger(), "Failed to open camera device %d", device_id);
      return;
    }

    cap_.set(cv::CAP_PROP_FRAME_WIDTH,  width_);
    cap_.set(cv::CAP_PROP_FRAME_HEIGHT, height_);
    cap_.set(cv::CAP_PROP_FPS,          fps);

    img_pub_ = create_publisher<sensor_msgs::msg::Image>(
      "/camera/forward/image_raw", 10);

    int interval_ms = 1000 / fps;
    timer_ = create_wall_timer(
      std::chrono::milliseconds(interval_ms),
      std::bind(&CameraNode::captureFrame, this));

    RCLCPP_INFO(get_logger(),
      "Camera node started: device=%d %dx%d @ %dfps",
      device_id, width_, height_, fps);
  }

private:
  void captureFrame() {
    cv::Mat frame;
    if (!cap_.read(frame) || frame.empty()) {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 5000,
        "Failed to capture frame");
      return;
    }

    auto msg = cv_bridge::CvImage(
      std_msgs::msg::Header(), "bgr8", frame).toImageMsg();
    msg->header.stamp    = now();
    msg->header.frame_id = frame_id_;
    img_pub_->publish(*msg);
  }

  cv::VideoCapture cap_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr img_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  int         width_    = 640;
  int         height_   = 480;
  std::string frame_id_ = "camera_forward_link";
};

int main(int argc, char ** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CameraNode>());
  rclcpp::shutdown();
}
